lib/Sema/CSGen.cpp

```diff
@@ -3667,35 +3667,54 @@ void ConstraintSystem::optimizeConstraints(Expr *e) {
   e->walk(optimizer);
 }
 
-bool swift::isExtensionApplied(DeclContext &DC, Type BaseTy,
-                               const ExtensionDecl *ED) {
-  if (!ED->isConstrainedExtension() ||
-      // We'll crash if we leak type variables from one constraint
+bool swift::isGenericContextApplied(DeclContext *DC, Type BaseTy,
+                                    GenericContext *GC) {
+  if (// We'll crash if we leak type variables from one constraint
       // system into the new one created below.
       BaseTy->hasTypeVariable() ||
       // We can't do anything if the base type has unbound generic
       // parameters either.
       BaseTy->hasUnboundGenericType())
     return true;
 
-  TypeChecker *TC = &createTypeChecker(DC.getASTContext());
-  TC->validateExtension(const_cast<ExtensionDecl *>(ED));
+  TypeChecker *TC = &createTypeChecker(DC->getASTContext());
+  DeclContext *baseDC = nullptr;
+  bool isExtension = false;
+
+  if (auto *ED = dyn_cast_or_null<ExtensionDecl>(GC->getAsDecl())) {
+    if (!ED->isConstrainedExtension())
+      return true;
+    TC->validateExtension(ED);
+    baseDC = ED->getExtendedNominal();
+    isExtension = true;
+  } else if (auto *VD = dyn_cast_or_null<ValueDecl>(GC->getAsDecl())) {
+    TC->validateDeclForNameLookup(VD);
+    if (!GC->getGenericSignature())
+      return true;
+    baseDC = VD->getDeclContext();
+  } else {
+    // Unknown generic context? Assume it applicable.
+    return true;
+  }
 
   ConstraintSystemOptions Options;
-  ConstraintSystem CS(*TC, &DC, Options);
+  ConstraintSystem CS(*TC, DC, Options);
   auto Loc = CS.getConstraintLocator(nullptr);
 
   // Prepare type substitution map.
-  SubstitutionMap Substitutions = BaseTy->getContextSubstitutionMap(
-    DC.getParentModule(), ED);
+  SubstitutionMap Substitutions =
+      BaseTy->getMetatypeInstanceType()->getContextSubstitutionMap(
```

Comment:
I would suggest doing the getMetatypeInstanceType() in the caller. It's simpler to maintain invariants if functions like this only take a small set of types, with various forms of desugaring done by the caller.